File: CattyTests/Project/ProjectTests.m
/**
 *  Copyright (C) 2010-2019 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

#import <XCTest/XCTest.h>
#import "Project.h"
#import "StartScript.h"
#import "IfThenLogicBeginBrick.h"
#import "IfThenLogicEndBrick.h"
#import "IfLogicBeginBrick.h"
#import "IfLogicElseBrick.h"
#import "IfLogicEndBrick.h"
#import "OrderedMapTable.h"
#import "SetVariableBrick.h"
#import "CBFileManager.h"
#import "AppDelegate.h"
#import "Util.h"
#import "LanguageTranslationDefines.h"

@interface ProjectTests : XCTestCase

@property (nonatomic, strong) Project *project;
@property (nonatomic, strong) CBFileManager *fileManager;

@end

@implementation ProjectTests

- (void)setUp
{
    [super setUp];
    if (! [self.fileManager directoryExists:[Project basePath]]) {
        [self.fileManager createDirectory:[Project basePath]];
    }
}

- (void)tearDown
{
    if (self.project) {
        [ProjectTests removeProject:[self.project projectPath]];
    }
    self.project = nil;
    self.fileManager = nil;
    [super tearDown];
}

- (void)setupForNewProject
{
    self.project = [Project defaultProjectWithName:kLocalizedNewProject projectID:nil];
}

- (void)testNewProjectIfProjectFolderExists
{
    [self setupForNewProject];
    XCTAssertTrue([self.fileManager directoryExists:[self.project projectPath]], @"No project folder created for the new project");
}

- (void)testNewProjectIfImagesFolderExists
{
    [self setupForNewProject];
    NSString *imagesDirName = [NSString stringWithFormat:@"%@%@", [self.project projectPath], kProjectImagesDirName];
    XCTAssertTrue([self.fileManager directoryExists:imagesDirName], @"No images folder created for the new project");
}

- (void)testNewProjectIfSoundsFolderExists
{
    [self setupForNewProject];
    NSString *soundsDirName = [NSString stringWithFormat:@"%@%@", [self.project projectPath], kProjectSoundsDirName];
    XCTAssertTrue([self.fileManager directoryExists:soundsDirName], @"No sounds folder created for the new project");
}

- (void)testCopyObjectWithIfThenLogicBeginBrick
{
    NSString *objectName = @"newObject";
    NSString *copiedObjectName = @"copiedObject";
    
    [self setupForNewProject];
    
    SpriteObject *object = [SpriteObject new];
    object.name = objectName;
    
    IfThenLogicBeginBrick *ifThenLogicBeginBrick = [IfThenLogicBeginBrick new];
    ifThenLogicBeginBrick.ifCondition = [[Formula alloc] initWithDouble:2];
    
    IfThenLogicEndBrick *ifThenLogicEndBrick = [IfThenLogicEndBrick new];
    ifThenLogicBeginBrick.ifEndBrick = ifThenLogicEndBrick;
    ifThenLogicEndBrick.ifBeginBrick = ifThenLogicBeginBrick;
    
    StartScript *script = [StartScript new];
    [script.brickList addObjectsFromArray:@[ifThenLogicBeginBrick, ifThenLogicEndBrick]];
    [object.scriptList addObject:script];
    [self.project.objectList addObject:object];
    
    NSUInteger initialObjectSize = self.project.objectList.count;
    
    SpriteObject *copiedObject = [self.project copyObject:object withNameForCopiedObject:copiedObjectName];
    XCTAssertEqual(1, copiedObject.scriptList.count);
    
    NSArray<SpriteObject*> *objectList = self.project.objectList;
    XCTAssertEqual(initialObjectSize + 1, objectList.count);
    XCTAssertTrue([objectList[initialObjectSize].name isEqualToString:copiedObjectName]);
    
    XCTAssertEqual(2, copiedObject.scriptList[0].brickList.count);
    XCTAssertTrue([copiedObject.scriptList[0].brickList[0] isKindOfClass:[IfThenLogicBeginBrick class]]);
    XCTAssertTrue([copiedObject.scriptList[0].brickList[1] isKindOfClass:[IfThenLogicEndBrick class]]);
    
    IfThenLogicBeginBrick *beginBrick = (IfThenLogicBeginBrick*) copiedObject.scriptList[0].brickList[0];
    IfThenLogicEndBrick *endBrick = (IfThenLogicEndBrick*) copiedObject.scriptList[0].brickList[1];
    
    XCTAssertEqual(endBrick, beginBrick.ifEndBrick);
    XCTAssertEqual(beginBrick, endBrick.ifBeginBrick);
    XCTAssertNotEqual(ifThenLogicEndBrick, beginBrick.ifEndBrick);
    XCTAssertNotEqual(ifThenLogicBeginBrick, endBrick.ifBeginBrick);
}

- (void)testCopyObjectWithIfTLogicBeginBrick
{
    NSString *objectName = @"newObject";
    NSString *copiedObjectName = @"copiedObject";
    
    [self setupForNewProject];
    
    SpriteObject *object = [SpriteObject new];
    object.name = objectName;
    
    IfLogicBeginBrick *ifLogicBeginBrick = [IfLogicBeginBrick new];
    ifLogicBeginBrick.ifCondition = [[Formula alloc] initWithDouble:1];
    
    IfLogicElseBrick *ifLogicElseBrick = [IfLogicElseBrick new];
    ifLogicElseBrick.ifBeginBrick = ifLogicBeginBrick;
    ifLogicBeginBrick.ifElseBrick = ifLogicElseBrick;
    
    IfLogicEndBrick *ifLogicEndBrick = [IfLogicEndBrick new];
    ifLogicBeginBrick.ifEndBrick = ifLogicEndBrick;
    ifLogicEndBrick.ifBeginBrick = ifLogicBeginBrick;
    ifLogicEndBrick.ifElseBrick = ifLogicElseBrick;
    
    StartScript *script = [StartScript new];
    [script.brickList addObjectsFromArray:@[ifLogicBeginBrick, ifLogicElseBrick, ifLogicEndBrick]];
    [object.scriptList addObject:script];
    [self.project.objectList addObject:object];
    
    NSUInteger initialObjectSize = self.project.objectList.count;
    
    SpriteObject *copiedObject = [self.project copyObject:object withNameForCopiedObject:copiedObjectName];
    XCTAssertEqual(1, copiedObject.scriptList.count);
    
    NSArray<SpriteObject*> *objectList = self.project.objectList;
    XCTAssertEqual(initialObjectSize + 1, objectList.count);
    XCTAssertTrue([objectList[initialObjectSize].name isEqualToString:copiedObjectName]);
    
    XCTAssertEqual(3, copiedObject.scriptList[0].brickList.count);
    XCTAssertTrue([copiedObject.scriptList[0].brickList[0] isKindOfClass:[IfLogicBeginBrick class]]);
    XCTAssertTrue([copiedObject.scriptList[0].brickList[1] isKindOfClass:[IfLogicElseBrick class]]);
    XCTAssertTrue([copiedObject.scriptList[0].brickList[2] isKindOfClass:[IfLogicEndBrick class]]);
    
    IfLogicBeginBrick *beginBrick = (IfLogicBeginBrick*) copiedObject.scriptList[0].brickList[0];
    IfLogicElseBrick *elseBrick = (IfLogicElseBrick*) copiedObject.scriptList[0].brickList[1];
    IfLogicEndBrick *endBrick = (IfLogicEndBrick*) copiedObject.scriptList[0].brickList[2];
    
    XCTAssertEqual(endBrick, beginBrick.ifEndBrick);
    XCTAssertEqual(elseBrick, beginBrick.ifElseBrick);
    XCTAssertEqual(elseBrick, endBrick.ifElseBrick);
    XCTAssertEqual(beginBrick, elseBrick.ifBeginBrick);
    XCTAssertEqual(endBrick, elseBrick.ifEndBrick);
    XCTAssertEqual(beginBrick, endBrick.ifBeginBrick);
    
    XCTAssertNotEqual(ifLogicEndBrick, beginBrick.ifEndBrick);
    XCTAssertNotEqual(ifLogicElseBrick, beginBrick.ifElseBrick);
    XCTAssertNotEqual(ifLogicElseBrick, endBrick.ifElseBrick);
    XCTAssertNotEqual(ifLogicBeginBrick, elseBrick.ifBeginBrick);
    XCTAssertNotEqual(ifLogicEndBrick, elseBrick.ifEndBrick);
    XCTAssertNotEqual(ifLogicBeginBrick, endBrick.ifBeginBrick);
}

- (void)testCopyObjectWithObjectVariable
{
    [self setupForNewProject];
    
    SpriteObject *object = [SpriteObject new];
    object.name = @"newObject";
    [self.project.objectList addObject:object];
    
    UserVariable *variable = [UserVariable new];
    variable.name = @"userVariable";
    [self.project.variables addObjectVariable:variable forObject:object];
    
    SetVariableBrick *setVariableBrick = [SetVariableBrick new];
    setVariableBrick.userVariable = variable;
    
    StartScript *script = [StartScript new];
    [script.brickList addObjectsFromArray:@[setVariableBrick]];
    [object.scriptList addObject:script];
    
    NSUInteger initialObjectSize = self.project.objectList.count;
    NSUInteger initialVariableSize = self.project.variables.allVariables.count;
    
    SpriteObject *copiedObject = [self.project copyObject:object withNameForCopiedObject:@"copiedObject"];
    XCTAssertEqual(1, copiedObject.scriptList.count);
    
    NSArray<SpriteObject*> *objectList = self.project.objectList;
    XCTAssertEqual(initialObjectSize + 1, objectList.count);
    XCTAssertEqual(initialVariableSize + 1, self.project.variables.allVariables.count);
    XCTAssertTrue([objectList[initialObjectSize].name isEqualToString:copiedObject.name]);
    
    XCTAssertEqual(1, copiedObject.scriptList[0].brickList.count);
    XCTAssertTrue([copiedObject.scriptList[0].brickList[0] isKindOfClass:[SetVariableBrick class]]);
    
    SetVariableBrick *copiedSetVariableBrick = (SetVariableBrick*) copiedObject.scriptList[0].brickList[0];
    XCTAssertNotNil(copiedSetVariableBrick.userVariable);
    XCTAssertNotEqual(variable, copiedSetVariableBrick.userVariable);
    XCTAssertTrue([variable.name isEqualToString:copiedSetVariableBrick.userVariable.name]);
}

- (void)testCopyObjectWithObjectList
{
    [self setupForNewProject];
    
    SpriteObject *object = [SpriteObject new];
    object.name = @"newObject";
    [self.project.objectList addObject:object];
    
    UserVariable *list = [UserVariable new];
    list.name = @"userList";
    list.isList = YES;
    [self.project.variables addObjectList:list forObject:object];
    
    SetVariableBrick *setVariableBrick = [SetVariableBrick new];
    setVariableBrick.userVariable = list;
    
    StartScript *script = [StartScript new];
    [script.brickList addObjectsFromArray:@[setVariableBrick]];
    [object.scriptList addObject:script];
    
    NSUInteger initialObjectSize = self.project.objectList.count;
    NSUInteger initialListSize = self.project.variables.allLists.count;
    
    SpriteObject *copiedObject = [self.project copyObject:object withNameForCopiedObject:@"copiedObject"];
    XCTAssertEqual(1, copiedObject.scriptList.count);
    
    NSArray<SpriteObject*> *objectList = self.project.objectList;
    XCTAssertEqual(initialObjectSize + 1, objectList.count);
    XCTAssertEqual(initialListSize + 1, self.project.variables.allLists.count);
    XCTAssertTrue([objectList[initialObjectSize].name isEqualToString:copiedObject.name]);
    
    XCTAssertEqual(1, copiedObject.scriptList[0].brickList.count);
    XCTAssertTrue([copiedObject.scriptList[0].brickList[0] isKindOfClass:[SetVariableBrick class]]);
    
    SetVariableBrick *copiedSetVariableBrick = (SetVariableBrick*) copiedObject.scriptList[0].brickList[0];
    XCTAssertNotNil(copiedSetVariableBrick.userVariable);
    XCTAssertNotEqual(list, copiedSetVariableBrick.userVariable);
    XCTAssertTrue([list.name isEqualToString:copiedSetVariableBrick.userVariable.name]);
}

#pragma mark - getters and setters
- (CBFileManager*)fileManager {
    return [CBFileManager sharedManager];
}

+ (void)removeProject:(NSString*)projectPath
{
    CBFileManager *fileManager = [CBFileManager sharedManager];
    if ([fileManager directoryExists:projectPath])
        [fileManager deleteDirectory:projectPath];
    [Util setLastProjectWithName:nil projectID:nil];
}

@end

File: src/CattyTests/Project/ProjectTests.swift
/**
 *  Copyright (C) 2010-2020 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

@testable import Pocket_Code

import Nimble
import XCTest

class ProjectTest: XCTestCase {

    var project: Project!
    var fileManager: CBFileManager!
    var object: SpriteObject!
    var scene: Scene!

    override func setUp() {
        super.setUp()

        fileManager = CBFileManager.shared()
        deleteAllProjectsAndCreateDefaultProject()

        self.scene = project.scene
        self.object = SpriteObject()
        self.object.scene = scene
    }

    override func tearDown() {
        let projectPath = project.projectPath()

        if fileManager.directoryExists(projectPath) {
            fileManager.deleteDirectory(projectPath)
        }

        super.tearDown()
    }

    private func deleteAllProjectsAndCreateDefaultProject() {
        for loadingInfo in Project.allProjectLoadingInfos() as! [ProjectLoadingInfo] {
            fileManager.deleteDirectory(loadingInfo.basePath!)
        }
        fileManager.addDefaultProjectToProjectsRootDirectoryIfNoProjectsExist()

        let loadingInfos = Project.allProjectLoadingInfos()
        XCTAssertEqual(1, loadingInfos.count)

        project = Project(loadingInfo: (loadingInfos.first as! ProjectLoadingInfo))
    }

    func testProjectDirectoryExists() {
        let directoryExists = fileManager.directoryExists(project.projectPath())
        XCTAssertTrue(directoryExists)
    }

    func testImagesDirectoryExists() {
        let imageDirectory = project.scene.imagesPath()
        let directoryExists = fileManager.directoryExists(imageDirectory)
        XCTAssertTrue(directoryExists)
    }

    func testSoundsDirectoryExists() {
        let soundsDirectory = project.scene.soundsPath()
        let directoryExists = fileManager.directoryExists(soundsDirectory)
        XCTAssertTrue(directoryExists)
    }

    func testCopyObjectWithIfThenLogicBeginBrick() {
        let objectName = "newObject"
        let copiedObjectName = "copiedObject"

        object.name = objectName

        let ifThenLogicBeginBrick = IfThenLogicBeginBrick()
        ifThenLogicBeginBrick.ifCondition = Formula(double: 2)

        let ifThenLogicEndBrick = IfThenLogicEndBrick()
        ifThenLogicBeginBrick.ifEndBrick = ifThenLogicEndBrick
        ifThenLogicEndBrick.ifBeginBrick = ifThenLogicBeginBrick

        let script = StartScript()
        script.brickList.addObjects(from: [ifThenLogicBeginBrick, ifThenLogicEndBrick] as [AnyObject])
        object.scriptList.add(script)
        project.scene.add(object: object)

        let initialObjectSize = project.scene.objects().count
        XCTAssertTrue(initialObjectSize > 0)

        let copiedObject = project.scene.copy(object, withNameForCopiedObject: copiedObjectName)!
        XCTAssertEqual(1, copiedObject.scriptList.count)

        let objectList = project.scene.objects()
        XCTAssertEqual(initialObjectSize + 1, objectList.count)
        XCTAssertEqual(copiedObjectName, (objectList[initialObjectSize] ).name)

        let copiedScript = copiedObject.scriptList[0] as! Script
        XCTAssertEqual(2, copiedScript.brickList.count)
        XCTAssertTrue(copiedScript.brickList[0] is IfThenLogicBeginBrick)
        XCTAssertTrue(copiedScript.brickList[1] is IfThenLogicEndBrick)

        let beginBrick = copiedScript.brickList[0] as! IfThenLogicBeginBrick
        let endBrick = copiedScript.brickList[1] as! IfThenLogicEndBrick

        XCTAssertEqual(endBrick, beginBrick.ifEndBrick)
        XCTAssertEqual(beginBrick, endBrick.ifBeginBrick)
        XCTAssertNotEqual(ifThenLogicEndBrick, beginBrick.ifEndBrick)
        XCTAssertNotEqual(ifThenLogicBeginBrick, endBrick.ifBeginBrick)
    }

    func testCopyObjectWithIfLogicBeginBrick() {
        let objectName = "newObject"
        let copiedObjectName = "copiedObject"

        object.name = objectName

        let ifLogicBeginBrick = IfLogicBeginBrick()
        ifLogicBeginBrick.ifCondition = Formula(double: 1)

        let ifLogicElseBrick = IfLogicElseBrick()
        ifLogicElseBrick.ifBeginBrick = ifLogicBeginBrick
        ifLogicBeginBrick.ifElseBrick = ifLogicElseBrick

        let ifLogicEndBrick = IfLogicEndBrick()
        ifLogicBeginBrick.ifEndBrick = ifLogicEndBrick
        ifLogicEndBrick.ifBeginBrick = ifLogicBeginBrick
        ifLogicEndBrick.ifElseBrick = ifLogicElseBrick

        let script = StartScript()
        script.brickList.addObjects(from: [ifLogicBeginBrick, ifLogicElseBrick, ifLogicEndBrick] as [AnyObject])
        object.scriptList.add(script)
        project.scene.add(object: object)

        let initialObjectSize = project.scene.objects().count
        XCTAssertTrue(initialObjectSize > 0)

        let copiedObject = project.scene.copy(object, withNameForCopiedObject: copiedObjectName)!
        XCTAssertEqual(1, copiedObject.scriptList.count)

        let objectList = project.scene.objects()
        XCTAssertEqual(initialObjectSize + 1, objectList.count)
        XCTAssertEqual(copiedObjectName, (objectList[initialObjectSize] ).name)

        let copiedScript = copiedObject.scriptList[0] as! Script
        XCTAssertEqual(3, copiedScript.brickList.count)
        XCTAssertTrue(copiedScript.brickList[0] is IfLogicBeginBrick)
        XCTAssertTrue(copiedScript.brickList[1] is IfLogicElseBrick)
        XCTAssertTrue(copiedScript.brickList[2] is IfLogicEndBrick)

        let beginBrick = copiedScript.brickList[0] as! IfLogicBeginBrick
        let elseBrick = copiedScript.brickList[1] as! IfLogicElseBrick
        let endBrick = copiedScript.brickList[2] as! IfLogicEndBrick

        XCTAssertEqual(endBrick, beginBrick.ifEndBrick)
        XCTAssertEqual(elseBrick, beginBrick.ifElseBrick)
        XCTAssertEqual(elseBrick, endBrick.ifElseBrick)
        XCTAssertEqual(beginBrick, elseBrick.ifBeginBrick)
        XCTAssertEqual(endBrick, elseBrick.ifEndBrick)
        XCTAssertEqual(beginBrick, endBrick.ifBeginBrick)

        XCTAssertNotEqual(ifLogicEndBrick, beginBrick.ifEndBrick)
        XCTAssertNotEqual(ifLogicElseBrick, beginBrick.ifElseBrick)
        XCTAssertNotEqual(ifLogicElseBrick, endBrick.ifElseBrick)
        XCTAssertNotEqual(ifLogicBeginBrick, elseBrick.ifBeginBrick)
        XCTAssertNotEqual(ifLogicEndBrick, elseBrick.ifEndBrick)
        XCTAssertNotEqual(ifLogicBeginBrick, endBrick.ifBeginBrick)
    }

    func testCopyObjectWithObjectVariable() {
        object.name = "newObjectName"
        project.scene.add(object: object)

        let variable = UserVariable(name: "userVariable")
        object.userData.add(variable)

        let setVariableBrick = SetVariableBrick()
        setVariableBrick.userVariable = variable

        let script = StartScript()
        script.brickList.addObjects(from: [setVariableBrick] as [AnyObject])
        object.scriptList.add(script)

        let initialObjectSize = project.scene.objects().count
        XCTAssertTrue(initialObjectSize > 0)

        let initialVariableSize = UserDataContainer.allVariables(for: project).count
        XCTAssertTrue(initialVariableSize > 0)

        let copiedObject = project.scene.copy(object, withNameForCopiedObject: "copiedObjectName")!
        XCTAssertEqual(1, copiedObject.scriptList.count)

        let objectList = project.scene.objects()
        XCTAssertEqual(initialObjectSize + 1, objectList.count)
        XCTAssertEqual(initialVariableSize + 1, UserDataContainer.allVariables(for: project).count)
        XCTAssertEqual((objectList[initialObjectSize] ).name, copiedObject.name)

        let copiedScript = copiedObject.scriptList[0] as! Script

        XCTAssertEqual(1, copiedScript.brickList.count)
        XCTAssertTrue(copiedScript.brickList[0] is SetVariableBrick)

        let copiedSetVariableBrick = copiedScript.brickList[0] as! SetVariableBrick
        XCTAssertNotNil(copiedSetVariableBrick.userVariable)
        XCTAssertFalse(variable === copiedSetVariableBrick.userVariable)
        XCTAssertEqual(variable.name, copiedSetVariableBrick.userVariable.name)
    }

    func testCopyObjectWithObjectList() {
        object.name = "newObjectName"
        project.scene.add(object: object)

        let list = UserList(name: "userList")
        object.userData.add(list)

        let brick = InsertItemIntoUserListBrick()
        brick.userList = list

        let script = StartScript()
        script.brickList.addObjects(from: [brick] as [AnyObject])
        object.scriptList.add(script)

        let initialObjectSize = project.scene.objects().count
        XCTAssertTrue(initialObjectSize > 0)

        let initialListSize = UserDataContainer.allLists(for: project).count
        XCTAssertTrue(initialListSize > 0)

        let copiedObject = project.scene.copy(object, withNameForCopiedObject: "copiedObjectName")!
        XCTAssertEqual(1, copiedObject.scriptList.count)

        let objectList = project.scene.objects()
        XCTAssertEqual(initialObjectSize + 1, objectList.count)
        XCTAssertEqual(initialListSize + 1, UserDataContainer.allLists(for: project).count)
        XCTAssertEqual((objectList[initialObjectSize] ).name, copiedObject.name)

        let copiedScript = copiedObject.scriptList[0] as! Script

        XCTAssertEqual(1, copiedScript.brickList.count)
        XCTAssertTrue(copiedScript.brickList[0] is InsertItemIntoUserListBrick)

        let copiedBrick = copiedScript.brickList[0] as! InsertItemIntoUserListBrick
        XCTAssertNotNil(copiedBrick.userList)
        XCTAssertFalse(list === copiedBrick.userList)
        XCTAssertEqual(list.name, copiedBrick.userList.name)
    }

    func testRenameProject() {
        let projectPath = project.projectPath()
        let projectId = project.header.programID

        XCTAssertTrue(fileManager.directoryExists(projectPath))

        project.rename(toProjectName: "newProject", andShowSaveNotification: true)

        let newProjectPath = project.projectPath()

        XCTAssertNotEqual(projectPath, newProjectPath)
        XCTAssertFalse(fileManager.directoryExists(projectPath))
        XCTAssertTrue(fileManager.directoryExists(newProjectPath))
        XCTAssertEqual(projectId, project.header.programID)
    }

    func testRenameProjectWithSameName() {
        let projectPath = project.projectPath()
        let projectId = project.header.programID

        XCTAssertTrue(fileManager.directoryExists(projectPath))

        project.rename(toProjectName: project.header.programName, andShowSaveNotification: true)

        let newProjectPath = project.projectPath()

        XCTAssertEqual(projectPath, newProjectPath)
        XCTAssertTrue(fileManager.directoryExists(newProjectPath))
        XCTAssertEqual(projectId, project.header.programID)
    }

    func testRenameProjectNameAndId() {
        let projectPath = project.projectPath()
        let newProjectId = "newProjectId"

        XCTAssertTrue(fileManager.directoryExists(projectPath))
        XCTAssertNotEqual(newProjectId, project.header.programID)

        project.rename(toProjectName: project.header.programName, andProjectId: newProjectId, andShowSaveNotification: true)

        let newProjectPath = project.projectPath()

        XCTAssertNotEqual(projectPath, newProjectPath)
        XCTAssertFalse(fileManager.directoryExists(projectPath))
        XCTAssertTrue(fileManager.directoryExists(newProjectPath))
        XCTAssertEqual(newProjectId, project.header.programID)
    }

    func testProjectWithLoadingInfoInvalidDirectory() {
        let loadingInfo = ProjectLoadingInfo.init(forProjectWithName: project.header.programName + "invalid", projectID: kNoProjectIDYetPlaceholder)

        XCTAssertFalse(fileManager.directoryExists(loadingInfo!.basePath))

        let project = Project(loadingInfo: loadingInfo!)
        XCTAssertNil(project)
    }

    func testProjectWithLoadingInfo() {
        let loadingInfo = ProjectLoadingInfo.init(forProjectWithName: project.header.programName, projectID: kNoProjectIDYetPlaceholder)

        XCTAssertTrue(fileManager.directoryExists(loadingInfo!.basePath))

        let project = Project(loadingInfo: loadingInfo!)
        XCTAssertNotNil(project)
        XCTAssertEqual(loadingInfo!.visibleName!, project!.header.programName!)
    }

    func testProjectWithLoadingInfoDivergingDirectoryAndProjectName() {
        let oldProjectName = project.header.programName
        let oldDirectoryName = project.projectPath()

        project.header.programName += "new"

        let newProjectName = project.header.programName
        XCTAssertNotEqual(oldProjectName, newProjectName)

        fileManager.moveExistingDirectory(atPath: oldDirectoryName, toPath: project.projectPath())

        let loadingInfo = ProjectLoadingInfo.init(forProjectWithName: newProjectName, projectID: kNoProjectIDYetPlaceholder)
        XCTAssertTrue(fileManager.directoryExists(loadingInfo!.basePath))

        let project = Project(loadingInfo: loadingInfo!)
        XCTAssertEqual(newProjectName, project!.header.programName!)
    }

    func testDefaultProjectWithoutSavedNotification() {
        let project = ProjectMock()

        project.rename(toProjectName: "Project Name", andShowSaveNotification: true)
        XCTAssertTrue(project.saveNotificationShown)

        project.translateDefaultProject()
        XCTAssertFalse(project.saveNotificationShown)
    }

    func testNotificationForValidProjectWithLoadingInfo() {
        let loadingInfo = ProjectLoadingInfo.init(forProjectWithName: project.header.programName, projectID: kNoProjectIDYetPlaceholder)

        expect(Project(loadingInfo: loadingInfo!)).to(postNotifications(equal([])))
    }

    func testNotificationForInvalidProjectWithLoading() {
        let loadingInfo = ProjectLoadingInfo.init(forProjectWithName: project.header.programName + "InvalidProject", projectID: kNoProjectIDYetPlaceholder)

        let expectedNotification = Notification(name: .projectInvalidVersion, object: loadingInfo)
        expect(Project(loadingInfo: loadingInfo!)).to(postNotifications(equal([expectedNotification])))
    }

    func testAllObject() {
        self.project = Project()
        self.project.scene = Scene()

        XCTAssertEqual(0, self.project.scene.objects().count)

        let objectA = SpriteObject()
        objectA.name = "objectA"

        project.scene.add(object: objectA)
        XCTAssertEqual(1, self.project.scene.objects().count)
        XCTAssertTrue(self.project.scene.objects()[0] === objectA)
    }

    func testChangeProjectOrientation() {
        let project = ProjectMock()
        let screenWidth = project.header.screenWidth
        let screenHeight = project.header.screenHeight

        project.header.landscapeMode = false
        project.changeOrientation()

        XCTAssertEqual(project.header.landscapeMode, true)
        XCTAssertEqual(screenWidth, project.header.screenHeight)
        XCTAssertEqual(screenHeight, project.header.screenWidth)
        project.changeOrientation()

        XCTAssertEqual(project.header.landscapeMode, false)
        XCTAssertEqual(screenWidth, project.header.screenWidth)
        XCTAssertEqual(screenHeight, project.header.screenHeight)
    }
}
